src/codeeditor/editor.ts

```diff
@@ -0,0 +1,804 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable
+} from 'phosphor/lib/core/disposable';
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  ISignal, clearSignalData, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  IChangedArgs
+} from '../common/interfaces';
+
+
+/**
+ * A namespace for code editors.
+ */
+export
+namespace CodeEditor {
+  /**
+   * A zero-based position in the editor.
+   */
+  export
+  interface ICursorPosition {
+    /**
+     * The cursor line number.
+     */
+    line: number;
+
+    /**
+     * The cursor column number.
+     */
+    column: number;
+  }
+
+  /**
+   * An interface describing editor state coordinates.
+   */
+  export
+  interface ICoords {
+    /**
+     * The left coordinate value.
+     */
+    readonly left: number;
+
+    /**
+     * The right coordinate value.
+     */
+    readonly right: number;
+
+    /**
+     * The top coordinate value.
+     */
+    readonly top: number;
+
+    /**
+     * The bottom coordinate value.
+     */
+    readonly bottom: number;
+  }
+
+  /**
+   * Configuration options for the editor.
+   */
+  export
+  interface IConfiguration extends IDisposable {
+    /**
+     * A signal emitted when the configuration changes.
+     */
+    changed: ISignal<IConfiguration, string>;
+
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly: boolean;
+
+    /**
+     * The number of spaces a tab is considered equal to. Defaults to 4.
+     */
+    tabSize: number;
+
+    /**
+     * Translate tabs to spaces. Defaults to true.
+     */
+    tabsToSpaces: boolean;
+
+    /**
+     * Set to false to stop auto pairing quotes, brackets etc.
+     * Defaults to true.
+     */
+    autoMatchEnabled: boolean;
+
+    /**
+     * Set to false to disable highlighting the brackets surrounding the
+     * cursor. Defaults to true.
+     */
+    matchBrackets: boolean;
+
+    /**
+     * Set to false to disable scrolling past the end of the buffer.
+     * Defaults to false.
+     */
+    scrollPastEnd: boolean;
+  }
+
+  /**
+   * A text selection.
+   */
+  export
+  interface ITextSelection {
+    /**
+     * The index to the first character in the current selection.
+     */
+    start: number;
+
+    /**
+     * The index to the last character in the current selection.
+     */
+    end: number;
+
+    /**
+     * The direction in which selection occurred.
+     */
+    direction: 'forward' | 'backward' | 'none';
```

Comment:
I see, Monaco also has a notion of a direction. Also it can be derived from the start and end positions for any concrete editor if we enforce the following contract:
- endPosition > startPosition => direction = 'forward'
- startPosition < endPosition => direction = 'backward'
- startPosition = endPosition => direction = 'none'

Could we add as a docstring?